packages/react-dom/src/client/ReactDOMHostConfig.js

```diff
@@ -1142,3 +1152,142 @@ export function getInstanceFromScope(
   }
   return null;
 }
+
+export const supportsTestSelectors = true;
+
+export function findFiberRoot(node: Instance): null | FiberRoot {
+  const stack = [node];
+  let index = 0;
+  while (index < stack.length) {
+    const current = stack[index++];
+    if (isContainerMarkedAsRoot(current)) {
+      return ((getInstanceFromNodeDOMTree(current): any): FiberRoot);
+    }
+    stack.push(...current.children);
+  }
+  return null;
+}
+
+export function getBoundingRect(node: Instance): BoundingRect {
+  const rect = node.getBoundingClientRect();
+  return {
+    x: rect.left,
+    y: rect.top,
+    width: rect.width,
+    height: rect.height,
+  };
+}
+
+export function matchAccessibilityRole(node: Instance, role: string): boolean {
+  if (role === getRole(node)) {
+    return true;
+  }
+
+  return false;
+}
+
+export function getTextContent(fiber: Fiber): string | null {
+  switch (fiber.tag) {
+    case HostComponent:
+      let textContent = '';
+      const childNodes = fiber.stateNode.childNodes;
+      for (let i = 0; i < childNodes.length; i++) {
+        const childNode = childNodes[i];
+        if (childNode.nodeType === Node.TEXT_NODE) {
+          textContent += childNode.textContent;
+        }
+      }
+      return textContent;
+    case HostText:
+      return fiber.stateNode.textContent;
+  }
+
+  return null;
+}
+
+export function isHiddenSubtree(fiber: Fiber): boolean {
+  return fiber.tag === HostComponent && fiber.memoizedProps.hidden === true;
+}
+
+export function setFocusIfFocusable(node: Instance): boolean {
+  // The logic for determining if an element is focusable is kind of complex,
+  // and since we want to actually change focus anyway- we can just skip it.
+  // Instead we'll just listen for a "focus" event to verify that focus was set.
+  //
+  // We could compare the node to document.activeElement after focus,
+  // but this would not handle the case where application code managed focus to automatically blur.
+  let didFocus = false;
+  const handleFocus = () => {
+    didFocus = true;
+  };
+
+  const element = ((node: any): HTMLElement);
+  try {
+    element.addEventListener('focus', handleFocus);
+    element.focus();
```

Comment:
This won't cover the case where you want to focus an SVG element. For that you can do:

```
(element.focus || HTMLElement.prototype.focus).call(element);
```
Another issue is that internally, we don't normally use `node.focus`, instead opting to use our `FocusManager` module, which sets the `tabIndex` to `-1` before calling `.focus()`. We do this because the focus handling logic internally might use a different heuristic from what is normally "browser focusable".

I'm also a bit concerned that this will trigger unwanted side-effects, i.e. calling `focus` would trigger an event listener that updates state or logs something. Which would be possible, right?

Reply:
> Another issue is that internally, we don't normally use `node.focus`, instead opting to use our `FocusManager` module, which sets the `tabIndex` to `-1` before calling `.focus()`. We do this because the focus handling logic internally might use a different heuristic from what is normally "browser focusable".

We want "normal browser focusable" behavior in this case, because code is for e2e testing. We want it to behave as closely to what the user would get as possible.

> I'm also a bit concerned that this will trigger unwanted side-effects, i.e. calling `focus` would trigger an event listener that updates state or logs something. Which would be possible, right?

This is for e2e testing. If focusing something triggers side effects, we want that behavior. This method won't be calling focus on more than one focusable thing.

Reply:
> We want "normal browser focusable" behavior in this case, because code is for e2e testing. We want it to behave as closely to what the user would get as possible.

Gotcha.

> This is for e2e testing. If focusing something triggers side effects, we want that behavior. This method won't be calling focus on more than one focusable thing.

In which case, we should also maybe provide an argument to `focusWithin` that tells it to scroll or not. This is an important heuristic and one that native `HTMLElement.prototype.focus` takes too (although only in the latest browsers). Given this is likely to be only be used by Chrome, which supports this feature, it sounds like we should also provide it?

Reply:
> > This is for e2e testing. If focusing something triggers side effects, we want that behavior. This method won't be calling focus on more than one focusable thing.
> 
> In which case, we should also maybe provide an argument to `focusWithin` that tells it to `preventScroll` or not. This is an important heuristic and one that native `HTMLElement.prototype.focus` takes too (although only in the latest browsers). Given this is likely to be only be used by Chrome, which supports this feature, it sounds like we should also provide it?

I'm not sure I understand how one follows from the other. Can you explain it to me?

This API is meant to let a test author *simulate* something the user would do with their mouse and/or keyboard. It's not meant as a replacement for programmatically calling `.focus()`.

Reply:
Now I'm confused. Why do you event want a `focusWithin` API at all then? I thought you wanted to manually invoke `focus` firing in e2e tests? Otherwise, if you're stating you only want to simulate user would do, then you never want to be doing this at all.

At a more fundamental level, there is focus with intent and focus without intent. Focusing with intent is when the user invokes it explicitly (clicking, tabbing, using arrows, VoiceOver etc) and without intent is when the focus is controlled by logic and usually imperatively handled (moving focus upon an animation completing, moving focus when something closes/opens etc).

If the API is meant to simulate what the user is doing, then you don't need this API. Simualting `click`, or `press tab` etc would achieve what you're trying to get here. In the caes you don't want to have focus with intent, then the UI logic should handle that.

Reply:
> Now I'm confused. Why do you event want a `focusWithin` API at all then?

This API would enable you to write a test that e.g. focuses inside of sign in form and then enters test account authentication details. For example, with Puppeteer that might look like:

```js
focusWithin(document.body, [createComponentSelector(LoginForm)]);

await page.keyboard.type("test-user");
```

> If the API is meant to simulate what the user is doing, then you don't need this API. Simulating `click`, or press `tab` etc would achieve what you're trying to get here.

I think this comment misses the high-level purpose of the selector API: It's to help you write tests that aren't brittle and that don't require custom test-specific data attributes (which have proven to be brittle and difficult to maintain at Facebook scale).

Reply:
> This API would enable you to write a test that e.g. focuses inside of sign in form and then enters test account authentication details. For example, with Puppeteer that might look like:

What field on the form would it select? Wouldn't you instead get the Username input and "click" it?

> Accomplishing this- without using the new React selector path to scope down which part of the page/application you want to interact with- would be difficult 😄

That's how our e2e tests look like in most cases already internally. I'm not sure you need to have this additionaly functionality. Otherwise, you might make tests that are fragile and break easily with small changes to the UI.

Reply:
> What field on the form would it select? Wouldn't you instead get the Username input and "click" it?

The first focusable one within the matched path, as described in the API docs in the PR description :smile: You probably wouldn't write the *actual exact syntax* I wrote. It was just an example.

> That's how our e2e tests look like in most cases already internally. I'm not sure you need to have this additionaly functionality. Otherwise, you might make tests that are fragile and break easily with small changes to the UI.

Check out the internal doc I sent you in work chat. These new APIs are meant to *fix* several of the problems with our internal tests.

Reply:
The internal doc makes more sense now. I see some of the things you want to tackle.